sentry-android-core/src/main/java/io/sentry/android/core/AndroidUserCache.java

```diff
@@ -0,0 +1,90 @@
+package io.sentry.android.core;
+
+import android.content.SharedPreferences;
+import io.sentry.core.IUserCache;
+import io.sentry.core.protocol.User;
+import io.sentry.core.util.Objects;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+// TODO: to think, this could be a SentryOptionsCache that could be expanded, does it make sense?
+
+/** the Android User Cache class. It caches the user data in the Android SharedPreferences */
+public final class AndroidUserCache implements IUserCache {
+
+  // TODO: only these 3 fields make sense to me, otherwise caching a raw json would be better, but I
+  // don't want to do that.
+  private static final String EMAIL = "email";
+  private static final String ID = "id";
+  private static final String USERNAME = "user_name";
+
+  private final @NotNull SharedPreferences sharedPreferences;
+  private final @NotNull SentryAndroidOptions options;
+  private final @Nullable String sentryDeviceId;
+
+  public AndroidUserCache(
+      final @NotNull SentryAndroidOptions options,
+      final @NotNull SharedPreferences sharedPreferences,
+      final @Nullable String sentryDeviceId) {
+    this.options = Objects.requireNonNull(options, "SentryAndroidOptions is required.");
+    this.sharedPreferences =
+        Objects.requireNonNull(sharedPreferences, "SharedPreferences is required.");
+    this.sentryDeviceId = sentryDeviceId;
+  }
+
+  //  TODO: I believe this makes sense only for global hub SDKs, but if not,
+  //   how do we cache multiple users (because we have multiple scopes)?
+
+  @Override
+  public void setUser(final @Nullable User user) {
+    if (options.isCacheUserForSessions()) {
```

Comment:
I think we could separate the logic of whether the user data should be saved or nor from the actual Android code. I would imagine two classes. One called something like `UserCache` that holds `SentryAndroidOptions` and `SentryDeviceId` and this class task to a simple interface like `UserPersistance`, that has an Android implementation an actually saves the user's data to the `SharedPreferences`. Then we have the SDK logic separated from persisting the data to Android. It would be easier to test and easier to extend. What do you think?

Reply:
yep, we could do that, just not sure how deep this abstraction should be, do we really see this user caching thing going forward with other SDKs? like java-backend, java-desktop etc... otherwise, it's just more classes for a single block of code, which is an `if`.
@bruno-garcia  opinions?

Reply:
It could be pulled out from the Android code but I believe it would only make sense then in desktop apps, right? Since this User in server apps are usually one per request and doesn't really make sense to cache.
Tbh we're not even sure this feature will even be used, it's just a work around to support `distinctId` for sessions based on user defined `user id`.

The proposal for `distinctId` has an option of not dealing with this at all (using only the generated id) which sounds like a good first version until we can get customer feedback to make a more educated decision.